### PR TITLE
docs: Add ARCHITECTURE.md

### PR DESCRIPTION
**Pull request by losvedir**
#### Summary of changes
**Asana Ticket:** [📝 architecture.md for realtime-signs](https://app.asana.com/0/281253957434160/1200644806185945)

Small document to provide a higher level overview of Realtime Signs for new contributors. 

#### Reviewer Checklist
- [ ] Meets ticket's acceptance criteria
- [ ] Any new or changed functions have typespecs
- [ ] Tests were added for any new functionality (don't just rely on Codecov)
- [ ] This branch was deployed to the staging environment and is currently running with no unexpected increase in warnings, and no errors or crashes (compare on Splunk: [staging](https://mbta.splunkcloud.com/en-US/app/search/search?q=search%20index%3Drealtime-signs-dev%20%22%5Berror%5D%22%20OR%20%22%5Bwarn%5D%22%20OR%20%22CRASH%22&display.page.search.mode=verbose&dispatch.sample_ratio=1&earliest=-4h%40m&latest=now&sid=1545840107.3874236) vs. [prod](https://mbta.splunkcloud.com/en-US/app/search/search?q=search%20index%3Drealtime-signs-prod%20%22%5Berror%5D%22%20OR%20%22%5Bwarn%5D%22%20OR%20%22CRASH%22&display.page.search.mode=verbose&dispatch.sample_ratio=1&earliest=-4h%40m&latest=now&sid=1545840137.3874305))

### CODE REVIEW

File: ARCHITECTURE.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# Architecture
+
+This document describes the high-level architecture of realtime_signs. The goal of this document is to provide a useful
+orientation to a new contributor to the codebase.
+
+## Application description
+
+Realtime Signs is the app that puts subway countdown predictions on the black and yellow Daktronics LED signs in and around stations. It does this by continually (about 1/sec) downloading GTFS-rt data and sending HTTP POST requests as necessary to the ARINC head end server.
```
Comment (digitalcora) [inline ARCHITECTURE.md:8]: Possible clarification to close the loop: _"...to the ARINC head-end server, which in turn sends requests to the Station Control Units that control the signs at each station."_

File: ARCHITECTURE.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+## Process-level architecture
+
+The smallest granularity with which we can address signs in a station is called the "zone". Each station can have up to six zones: north, south, east, west, center, and mezzanine. Each zone throughout the system has its own dedicated Elixir process, a `Signs.Realtime` GenServer configured at start-up via `priv/signs.json`. See `Signs.Utilities.SourceConfig` for documentation of the format of the `signs.json` file.
```
Comment (digitalcora) [inline ARCHITECTURE.md:12]: > Each station can have up to six zones: north, south, east, west, center, and mezzanine.

I'd maybe clarify this a bit, though I suppose this knowledge isn't necessary to understand the app:

> Each station can have up to six zones, which are always named "north", "south", "east", "west", "center", and "mezzanine", though these are arbitrary identifiers and may not reflect the real layout of the station.

File: ARCHITECTURE.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+There are several "Engine" GenServers, responsible for maintaining various bits of data: alerts, predictions, signs-ui configuration, etc. Each sign queries the engines once per tick (roughly 1/sec) to see if its content needs to change.
```
Comment (digitalcora) [inline ARCHITECTURE.md:14]: > Each sign queries the engines

Does this mean "each `Signs.Realtime` process queries the engines"?

File: ARCHITECTURE.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+To prevent DOSing the ARINC head-end server, each sign does not send HTTP requests to it directly. Rather, there's a `MessageQueue` GenServer that the signs use to add their messages to. Separately, there are a number of `PaEss.HttpUpdater` GenServers that pull from the queue to send the messages. The number of updaters and the frequency with which they pull from the queue is fixed, in order to rate limit the requests to the head end server. The size of the MessageQueue is fixed to allow some buffering for hiccups in the network, while not being so large that very out of date messages end up getting sent to ARINC. The ARINC head-end server has a lot of cores, but is somewhat slow per-core, so we ended up with a largish number of updaters, each posting somewhat slowly.
```
Comment (digitalcora) [inline ARCHITECTURE.md:16]: > each sign does not send HTTP requests to it directly

> that the signs use to add their messages to

Same comment as above, basically. It seems like we're saying "a sign" when we mean "a zone" throughout this doc, especially in the "utility modules" section. We could say that this is the convention we use, but in that case it'd help to call it out explicitly at the beginning.

File: ARCHITECTURE.md [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+![Realtime Signs process diagram](/realtime_signs_processes.png)
```
Comment (digitalcora) [inline ARCHITECTURE.md:18]: I'm unclear on what the directions of the arrows in this diagram mean. Could be useful to have a "key"?

File: ARCHITECTURE.md [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+## Code organization
+
+The various "engines" which keep track of state that signs are named `Engine.*`.
```
Comment (digitalcora) [inline ARCHITECTURE.md:22]: > which keep track of state that signs

There seems to be a word or potentially several words missing here.

File: ARCHITECTURE.md [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+The main sign module is `Signs.Realtime` which is the GenServer responsible for keeping track of what's on a sign's two lines and when to play various audio announcements.
+
+Ultimately, the point of the app is to display text content on signs, and play equivalent audio. This is represented in the app via `Content`. Every kind of text that can be put onto the signs is a struct type of `Content.Message.*`. Every message can be converted into a type of `Content.Audio.*`. Both of these can be "serialized" by the app into requests to ARINC.
```
Comment (digitalcora) [inline ARCHITECTURE.md:26]: > is a struct type of `Content.Message.*`

> can be converted into a type of `Content.Audio.*`

Maybe clearer: "has a corresponding `Content.Message.*` struct" and "can be converted into a corresponding `Content.Audio.*` struct"

File: lib/signs/utilities/source_config.ex [hunk 8]
```diff
@@ -1,29 +1,35 @@
...
+    "pa_ess_loc": "OROX",
+    "text_zone": "s",
+    "audio_zones": ["s"],
+    "read_loop_offset": 60,
+    "source_config": [...]
+  },
+
+  * id: the internal ID of this "sign". Really, it's a "zone" and could have a few physical signs as a part of it in the station, but they all display exactly the same content. Logically they're a single entity in the ARINC system as far as we can interact with it. This ID is also used by signs-ui, and is how realtime-signs knows how signs-ui has configured the sign.
+  * headway_group: this references the signs-ui values that the PIOs set for headways.
+  * type: formerly there were different kinds of signs in realtime_signs, but now they're all "realtime" only.
```
Comment (digitalcora) [inline lib/signs/utilities/source_config.ex:20]: I'd lead with the important info:

```
* type: always "realtime". (there were other types in previous iterations of the app)
```

File: lib/signs/utilities/source_config.ex [hunk 9]
```diff
@@ -1,29 +1,35 @@
...
+  * audio_zones: a list of those ARINC zones for the purpose of sending audio. This may differ from the text_zone when the speakers are close enough to cause confusion. For example, at Park Street for the Red line, we have the center platform text_zone set to "c", but the audio_zones set to [], while the north zone we have the text_zone set to "n" and the audio_zones set to ["n", "c"]. So the center platform doesn't play audio of its *own*, but rather the north and south platforms play their audio over the center platform speakers. Since a train that goes to "ARR" will do so simultaneously for either the north or south platform *and* the center platform, this configuration prevents simultaneous, slightly overlapping audio of the same content.
+  * read_loop_offset: how many seconds to wait after app start-up before entering the "read loop", which reads things like "The next train to X arrives in 3 minutes. The following train arrives in 8 minutes". We generally have different read_loop_offsets for different zones, or where speakers are particularly close, to encourage them to play their audio at different times.
+  * source config: see below.
```
Comment (digitalcora) [inline lib/signs/utilities/source_config.ex:25]: `source config` => `source_config`

File: lib/signs/utilities/source_config.ex [hunk 10]
```diff
@@ -1,29 +1,35 @@
...
+
+  ## Source config
+  A sign's data is provided via the "source_config" key. This data is a list of lists of data sources, called "sources".
```
Comment (digitalcora) [inline lib/signs/utilities/source_config.ex:28]: More concisely, and explaining where this is defined:

```
A sign's data is provided via the "source_config" key,
which is a list of lists of "sources" (see next section).
```

File: lib/signs/utilities/source_config.ex [hunk 11]
```diff
@@ -1,29 +1,35 @@
...
+
+  A list of sources can be provided: [{...}, {...}, {...}, ...]. All the sources will be sorted together in order of their respective arrivals or departures (depending on their respective "terminal" values). For example, JFK mezzanine northbound uses 70086 and 70096 in a source group, since those are the Ashmont and Braintree platforms, and it wishes to display the next train from either location.
```
Comment (digitalcora) [inline lib/signs/utilities/source_config.ex:30]: I'm not sure what "sorted together" means here, or the implications of sources being sorted vs. not; maybe we could explain it?

File: lib/signs/utilities/source_config.ex [hunk 12]
```diff
@@ -37,6 +43,30 @@ defmodule Signs.Utilities.SourceConfig do
...
+    "terminal": false,
+    "announce_arriving": true,
+    "announce_boarding": false
+  }
+
+  * stop_id: The GTFS stop_id that it uses for prediction data.
+  * routes: A list of routes that are relevant to this sign regarding alerts.
+  * direction_id: 0 or 1, used in tandem with the stop ID for predictions
+  * headway_direction_name: The headsign to use when in "headway mode"
+  * platform: null or ashmont/braintree for JFK/UMass
```
Comment (digitalcora) [inline lib/signs/utilities/source_config.ex:66]: Can we explain what this controls / why it's here?

**Review by digitalcora — COMMENTED** [review]
Yay architecture! Good to have this documented. I'm learning things 🙂

General comment: can we hard-wrap these? At least the Elixir comments should be wrapped at the same line length as code (98).

File: ARCHITECTURE.md (hunk 5 above)
Comment (losvedir) [reply ARCHITECTURE.md:18]: They mean "sends a message to", but actually there's an error in that the Predictions Engine does not send messages to the individual GenServers. I stole this .png from my GenServer talk, but the original was on my last laptop. I used Preview.app to try to quickly edit out the wrong arrow heads and add a legend.

File: lib/signs/utilities/source_config.ex [hunk 13]
```diff
@@ -1,29 +1,57 @@
...
-  A list of sources can be provided in a group: [{...}, {...}, {...}, ...]. All the sources will be sorted
-  together in order of their respective arrivals/departures (depending on their respective "terminal" values).
-  For example, JFK mezzanine northbound uses 70086 and 70096 in a source group, since those are the Ashmont
-  and Braintree platforms, and it wishes to display the next train from either location.
-
-  We currently support up to two _lists_ of sources. If one list is provided, then its next two trips will
-  show up on the lines of the sign. If two lists are provided, then the next trip from each list will
-  appear on different lines of the sign.
+    "id": "roxbury_crossing_southbound", "headway_group": "orange_trunk", "type": "realtime",
+    "pa_ess_loc": "OROX", "text_zone": "s", "audio_zones": ["s"], "read_loop_offset": 60,
```
Comment (digitalcora) [inline lib/signs/utilities/source_config.ex:9]: Wrapping mishap? I assume these were supposed to stay one-key-per-line.

File: lib/signs/utilities/source_config.ex [hunk 14]
```diff
@@ -37,6 +65,32 @@ defmodule Signs.Utilities.SourceConfig do
...
   ]
+
+  ## Source
+
+  Allows specifying one of a sign's data sources that it uses to calculate what to display. It
+  looks like:
+
+  {
+    "stop_id": "70008", "routes": ["Orange"], "direction_id": 0, "headway_direction_name": "Forest
+    Hills", "platform": null, "terminal": false, "announce_arriving": true, "announce_boarding":
```
Comment (digitalcora) [inline lib/signs/utilities/source_config.ex:76]: As above re: wrapping

**Review by digitalcora — APPROVED** [review]
One small comment, looks good to me otherwise!

**Pull request merged**